lib/appmap/hook/method.rb

```diff
@@ -69,28 +69,24 @@ def activate
         after_hook = self.method(:after_hook)
         with_disabled_hook = self.method(:with_disabled_hook)
 
-        hook_method_def = Proc.new do |*args, &block|
-          is_array_containing_empty_hash = ->(obj) {
-            obj.is_a?(Array) && obj.length == 1 && obj[0].is_a?(Hash) && obj[0].size == 0
-          }
-
-          call_instance_method = -> {
-            # https://github.com/applandinc/appmap-ruby/issues/153
-            if NEW_RUBY && is_array_containing_empty_hash.(args) && hook_method.arity == 1
-              if NEW_RUBY
-                hook_method.bind_call(self, {}, &block)
-              else
-                hook_method.bind(self).call({}, &block)
-              end
+        is_array_containing_empty_hash = ->(obj) {
+          obj.is_a?(Array) && obj.length == 1 && obj[0].is_a?(Hash) && obj[0].size == 0
+        }
+
+        call_instance_method = lambda do |receiver, args, &block|
+          # https://github.com/applandinc/appmap-ruby/issues/153
+          if NEW_RUBY && is_array_containing_empty_hash.(args) && hook_method.arity == 1
+            hook_method.bind_call(receiver, {}, &block)
+          else
+            if NEW_RUBY
+              hook_method.bind_call(receiver, *args, &block)
             else
-              if NEW_RUBY
-                hook_method.bind_call(self, *args, &block)
-              else
-                hook_method.bind(self).call(*args, &block)
-              end
+              hook_method.bind(receiver).call(*args, &block)
             end
-          }
+          end
+        end
```

Comment:
Might make sense to try to skip the NEW_RUBY branch on every call:
```ruby
if NEW_RUBY
  call_instance_method = lambda do |receiver, args, &block|
    if args == [{}] && hook_method.arity == 1
      hook_method.bind_call(receiver, {}, &block)
    else
      hook_method.bind_call(receiver, *args, block)
    end
  end
else
  call_instance_method = lambda do |receiver, args, &block|
    hook_method.bind(receiver).call(*args, &block)
  end
end
```

BTW, is this [{}] workaround still needed for ruby 3?